### PR TITLE
List map and key value map

### PR DESCRIPTION
**Pull request by YouriMulder**

### CODE REVIEW

**Review by Jaluakbar — COMMENTED** [review]
Code look clean, only thing i dont understand is the class PointState, but thats on me

**Review by Jaluakbar — APPROVED** [review]

**Comment by YouriMulder** [comment]
I will add some doxygen for the PointState enum.

File: list_map.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        return points
+
+    def is_valid_coordinate(self, coordinate):
+        """
+        @brief Method used to check if the coordinate is valid.
+        @return A boolean containing True if valid else False.
+        """
+
+        if coordinate.y > len(self.map) - 1 or coordinate.x > len(self.map[coordinate.y]) - 1:
+            return False
```
Comment (bartbal) [inline list_map.py]: what will it return when the if statement is false?

File: polar_coordinate.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#!/bin/python3
+
+from math import sin, cos, pi
+from enum import Enum
+import time
+
```
Comment (bartbal) [inline polar_coordinate.py:6]: double white lines around classes

File: quadtree_map.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
+from random import randint
+from cartesian_coordinate import CartesianCoordinate
+from mapping_interface import MapInterface
+from mapping_interface import PointState
+
```
Comment (bartbal) [inline quadtree_map.py:5]: double white lines around classes

File: quadtree_map.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                point.x <= self.x + self.width and
+                point.y >= self.y - self.height and
+                point.y <= self.y + self.height)
+
+    def intersects(self, other):
+        """
+        @brief A function if a another rectangle intersects with this rectangle.
+        @return Boolean.
+        """
+        return not((other.x - other.width > self.x + self.width) or
```
Comment (bartbal) [inline quadtree_map.py]: operators before expression

File: quadtree_map.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+        return self.topleft.add_point_cartesian(point, state, depth+1) \
+                or self.topright.add_point_cartesian(point, state, depth+1) \
+                or self.bottomleft.add_point_cartesian(point, state, depth+1) \
+                or self.bottomright.add_point_cartesian(point, state, depth+1)
+
+    def expand(self):
+        """
+        @brief Expand function used to make the map bigger when a new point is added that is outside the current size of the map, makes the map 2x bigger.
+        """
```
Comment (bartbal) [inline quadtree_map.py:102]: add white line

File: quadtree_map.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+        self.divided = False
+        self.points.clear()
+
+        for p in all_points:
+            self.add_point_cartesian(p, PointState.unoccupied)
+
+
+    def delete(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:117]: add white line

File: quadtree_map.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+            self.bottomleft.delete()
+            self.bottomright.delete()
+
+            del self.topleft
+            del self.topright
+            del self.bottomleft
+            del self.bottomright
+
+    def subdivide(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:133]: add white line

File: quadtree_map.py [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+        br = rectangle(x+w, y+h, w, h)
+        self.bottomright = quadtreeMap(br, self.capacity)
+
+        bl = rectangle(x-w, y+h, w, h)
+        self.bottomleft = quadtreeMap(bl, self.capacity)
+
+        self.divided = True
+
+    def query(self, selected_range, found):
+        """
```
Comment (bartbal) [inline quadtree_map.py:157]: add white line

File: quadtree_map.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+                found.append(p)
+
+        if self.divided:
+            self.topleft.query(selected_range, found)
+            self.topright.query(selected_range, found)
+            self.bottomleft.query(selected_range, found)
+            self.bottomright.query(selected_range, found)
+
+    def get_map_points(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:177]: add white line

File: quadtree_map.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+        @brief A function that returns all the points currently in the map.
+        @return A list of all the points found.
+        """
+        points = []
+        self.get_points(points)
+        return points
+
+    def get_points(self, all_points):
+        """
```
Comment (bartbal) [inline quadtree_map.py:187]: add white line

File: quadtree_map.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+            all_points.append(p)
+
+        if self.divided:
+            self.topleft.get_points(all_points)
+            self.topright.get_points(all_points)
+            self.bottomleft.get_points(all_points)
+            self.bottomright.get_points(all_points)
+
+    def print_tree(self, rootnumber=0, treename="root"):
+        """
```
Comment (bartbal) [inline quadtree_map.py:202]: add white line

File: quadtree_map.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        if len(self.points) > 0:
+            print("Rootdepth = ", rootnumber,
+                  ", position = ", treename,
+                  ", points = ", len(self.points))
+            for p in self.points:
+                print(p)
+        else:
+            return
+
```
Comment (bartbal) [inline quadtree_map.py:214]: useless else

File: quadtree_map.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+        if self.divided:
+            self.topleft.print_tree(rootnumber + 1, treename + " topleft")
+            self.topright.print_tree(rootnumber + 1, treename + " topright")
+            self.bottomleft.print_tree(rootnumber + 1, treename + " bottomleft")
+            self.bottomright.print_tree(rootnumber + 1, treename + " bottomright")
+        print()
+
+    def get_top_left_point(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:226]: add white line

File: quadtree_map.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+        @brief Method used to get the topLeftPoint of a map.
+        @return A CartesianCoordinate containing the top left corner point of a map.
+        """
+        return CartesianCoordinate(self.boundary.x - self.boundary.width, self.boundary.y - self.boundary.height)
+
+    def get_bot_right_point(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:234]: add white line

File: quadtree_map.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+        @brief Method used to get the height of a map.
+        @return A integer containing the height of a map.
+        """
+
+        return (self.boundary.y + self.boundary.height) - (self.boundary.y - self.boundary.height)
+
+
+    def get_width(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:250]: add white line

File: quadtree_map.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+        @brief Method used to get the width of a map.
+        @return A integer containing the width of a map.
+        """
+        return (self.boundary.x + self.boundary.width) - (self.boundary.x - self.boundary.width)
+
+    def print_map(self):
+        """
```
Comment (bartbal) [inline quadtree_map.py:258]: add white line

File: quadtree_map.py [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+        all_points = self.get_map_points()
+
+        for x in range(x_start, x_end):
+            for y in range(y_start, y_end):
+                if CartesianCoordinate(x, y) in all_points:
+                    print("#", end="")
+                else:
+                    print("-", end="")
+            print()
+        return
```
Comment (bartbal) [inline quadtree_map.py]: useless return

File: quadtree_map.py [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+        y_end = selected_range.y + selected_range.height
+
+        for x in range(x_start, x_end):
+            for y in range(y_start, y_end):
+                if CartesianCoordinate(x, y) in query_points:
+                    print("#", end="")
+                else:
+                    print("-", end="")
+            print()
+        return
```
Comment (bartbal) [inline quadtree_map.py]: useless return

File: quadtree_map.py [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+                if CartesianCoordinate(x, y) in all_points:
+                    print("#", end="")
+                else:
+                    print("-", end="")
+            print()
+        return
+
+
+    def print_query(self, selected_range, query_points):
+        """
```
Comment (bartbal) [inline quadtree_map.py:280]: add white line

File: quadtree_map.py [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+                if CartesianCoordinate(x, y) in query_points:
+                    print("#", end="")
+                else:
+                    print("-", end="")
+            print()
+        return
+
+
+    def isOccupied(self, coordinate):
+        """
```
Comment (bartbal) [inline quadtree_map.py:298]: add white line

File: quadtree_map.py [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+        @param boundary The quadtree size/boundary(a rectangle class).
+        @param capacity The maximal capacity of the tree node before it has to subdivide.
+        """
+        self.boundary = boundary
+        self.capacity = capacity
+        self.points = []
+        self.divided = False
+
+    def add_point_cartesian(self, point, state, depth=0):
+        """
```
Comment (bartbal) [inline quadtree_map.py:74]: add white line

File: quadtree_map.py [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+                   (other.x + other.width < self.x - self.width) or
+                   (other.y - other.height > self.y + self.height) or
+                   (other.y + other.height < self.y - self.height))
+
+class quadtreeMap(MapInterface):
+    """
+    @brief A tree node that can divide in 4 smaller nodes. First is root.
+    """
+    def __init__(self, boundary, capacity):
+        """
```
Comment (bartbal) [inline quadtree_map.py:62]: add white line

File: quadtree_map.py [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+        @param w The width of the rectangle(left and right from the middle point).
+        @param h The height of the rectangle(above and below from the middle point).
+        """
+        self.x = x
+        self.y = y
+        self.width = w
+        self.height = h
+
+    def contains(self, point):
+        """
```
Comment (bartbal) [inline quadtree_map.py:32]: add white line

File: quadtree_map.py [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+from cartesian_coordinate import CartesianCoordinate
+from mapping_interface import MapInterface
+from mapping_interface import PointState
+
+"""
+@package Quadtree
+This module is the implementation of mapping using quadtree algorithm.
+"""
+class rectangle():
+    """
```
Comment (bartbal) [inline quadtree_map.py:13]: add white line

**Review by bartbal — CHANGES_REQUESTED** [review]

**Review by bartbal — APPROVED** [review]

**Pull request merged**